network/wsNetwork.go

```diff
@@ -843,15 +875,39 @@ func (wn *WebsocketNetwork) ServeHTTP(response http.ResponseWriter, request *htt
 		return
 	}
 
+	// check to see that the genesisID in the request URI is valid and matches the supported one.
+	pathVars := mux.Vars(request)
+	otherGenesisID, hasGenesisID := pathVars["genesisID"]
+	if !hasGenesisID || otherGenesisID == "" {
+		response.WriteHeader(http.StatusNotFound)
+		return
+	}
+	if wn.GenesisID != otherGenesisID {
+		wn.log.Warnf("new peer %#v genesis mismatch, mine=%#v theirs=%#v, headers %#v", request.RemoteAddr, wn.GenesisID, otherGenesisID, request.Header)
+		response.WriteHeader(http.StatusPreconditionFailed)
+		response.Write([]byte("mismatching genesis ID"))
+		return
+	}
+
 	// TODO: rate limit incoming connections. (must wait at least Duration between disconnect and connect? no more than N connect attempts per Duration?)
-	wn.log.Debugf("inbound from %s", request.RemoteAddr)
-	ok, otherTelemetryGUID, otherPublicAddr, otherInstanceName := wn.checkHeaders(request.Header, request.RemoteAddr, originIP)
-	if !ok {
-		networkConnectionsDroppedTotal.Inc(map[string]string{"reason": "bad header"})
+	otherTelemetryGUID, otherPublicAddr, otherInstanceName, checkErr := wn.checkHeaders(request.Header, request.RemoteAddr, originIP)
+	if checkErr != nil {
+		networkConnectionsDroppedTotal.Inc(map[string]string{"reason": checkErr.Type()})
+		switch checkErr.(type) {
+		case *matchingRandomIDError:
+			response.WriteHeader(http.StatusLoopDetected)
+		default:
+			response.WriteHeader(http.StatusPreconditionFailed)
+		}
+		n, err := response.Write([]byte(checkErr.Error()))
+		if err != nil {
+			wn.log.Warnf("ws failed to write response '%s' : n = %d err = %v", checkErr.Error(), n, err)
+		}
 		return
 	}
 	requestHeader := make(http.Header)
 	wn.setHeaders(requestHeader)
+	requestHeader.Set(GenesisHeader, wn.GenesisID)
```

Comment:
This is already done in wn.setHeaders()

Reply:
it was removed from the setHeaders, since the setHeaders only sets the common headers.
This header is used only on the response now to avoid conflicting with the genesis-id passed as past of the url.

Reply:
ok thanks - I didn't pick up that change.